charmcraft/commands/store/registry.py

```diff
@@ -193,6 +198,63 @@ def upload_manifest(self, manifest_data, reference):
         assert_response_ok(response, expected_status=201)
         logger.debug("Manifest uploaded OK")
 
+    def upload_blob(self, filepath, size, digest):
+        """Upload the blob from a file."""
+        # get the first URL to start pushing the blob
+        logger.debug("Getting URL to push the blob")
+        url = self._get_url("blobs/uploads/")
+        response = self._hit("POST", url)
+        assert_response_ok(response, expected_status=202)
+        upload_url = response.headers["Location"]
+        range_from, range_to_inclusive = [
+            int(x) for x in response.headers["Range"].split("-")
+        ]
+        logger.debug(
+            "Got upload URL ok with range %s-%s", range_from, range_to_inclusive
+        )
+        if range_from != 0:
+            raise CommandError("Server error: bad range received")
+        if range_to_inclusive == 0:
+            range_to_inclusive = -1
+
+        # start the chunked upload
+        from_position = range_to_inclusive + 1
```

Comment:
This looks a bit suspicious, is it correct that we'll never start at position 1? Maybe positions are always a multiple of `CHUNK_SIZE` and therefore always greater than 1? If so, I think it's worth to add a comment explaining the situation.

Reply:
Oh, it seems that @sergiusens found the comment in line 42.

Reply:
It's a side effect of the range being inclusive.

So, if the server tells us that it already has "0-80", means that it has 81 bytes (from 0 to 80 inclusive), we set `from_position` in 81 and read from there. Going down, if the server says it has "0-10" it means it has 11 bytes. "0-2" would mean it has 3 bytes (from positions 0, 1 and 2). "0-1" would mean it has bytes 0 and 1. "0-0" would mean... WAIT! this is what the store returns when it does not have ANY bytes at all.

So we comply with Range parameter, but addressing this special case, I don't see many alternatives here. The worst think it could happen, IIUC, is that we start from 0 when the server has 1 byte already, which is not a problem.

Reply:
this could be a code comment itself. I remember this comment from a past discussion and quick demo you did.

Reply:
Added a comment explaining this...